scripts/nb-tester/qiskit_docs_notebook_tester/execute.py

```diff
@@ -0,0 +1,140 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may obtain a
+# copy of this license in the LICENSE file in the root directory of this source
+# tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this copyright
+# notice, and modified files need to carry a notice indicating that they have
+# been altered from the originals.
+
+from __future__ import annotations
+
+import tempfile
+import textwrap
+from dataclasses import dataclass
+
+import nbclient
+import nbformat
+from jupyter_client.manager import start_new_async_kernel, AsyncKernelClient
+
+from .config import NotebookJob, Result
+
+
+@dataclass(frozen=True)
+class NotebookWarning:
+    cell_index: int
+    msg: str
+
+    def format(self) -> str:
+        """
+        Format warning to pretty string
+        """
+        message = f"Cell {self.cell_index}:\n"
+        for line in self.msg.splitlines():
+            message += (
+                textwrap.fill(
+                    line, width=77, initial_indent=" │ ", subsequent_indent=" │ "
+                )
+                + "\n"
+            )
+        return message
+
+
+def extract_warnings(notebook: nbformat.NotebookNode) -> list[NotebookWarning]:
+    """
+    Detect warning messages in cell outputs
+    """
+    notebook_warnings = []
+    for cell_index, cell in enumerate(notebook.cells):
+        if not hasattr(cell, "outputs"):
+            continue
+        if "ignore-warnings" in cell.metadata.get("tags", []):
+            continue
+        for output in cell.outputs:
+            if hasattr(output, "name") and output.name == "stderr":
+                notebook_warnings.append(
+                    NotebookWarning(cell_index=cell_index, msg=output.text)
+                )
+    return notebook_warnings
+
+
+async def execute_notebook(job: NotebookJob) -> Result:
```

Comment:
This function is logically the same, but it gets its information from the `NotebookJob` rather than the separate arguments.